### PR TITLE
Oauth support

### PR DESCRIPTION
**Pull request by clayrosenthal**
# Proposed Changes

Adding support for oauth client authentication. See [Tailscale docs](https://tailscale.com/kb/1215/oauth-clients/)

Probably wanna squash this merge, it has commits from my original PR  #352

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -144,7 +144,7 @@ repos:
         name: 🆎 Static type checking using pyright
         language: system
         types: [python]
-        entry: npm run pyright
+        entry: npm run pyright --venvpath $(poetry env info -p)
```
Comment (frenck) [inline .pre-commit-config.yaml:147]: Unrelated change

File: README.md [hunk 2]
```diff
@@ -71,7 +71,7 @@ This is an active open-source project. We are always open to people who want to
 use the code or contribute to it.
 
 We've set up a separate document for our
-[contribution guidelines](CONTRIBUTING.md).
+[contribution guidelines](.github/CONTRIBUTING.md).
```
Comment (frenck) [inline README.md:74]: Unrelated change

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "tailscale",
-  "version": "0.0.0",
+  "version": "0.4.0",
```
Comment (frenck) [inline package.json:3]: This should not be changed

File: pyproject.toml [hunk 4]
```diff
@@ -1,8 +1,8 @@
 [tool.poetry]
 name = "tailscale"
-version = "0.0.0"
+version = "0.4.0"
 description = "Asynchronous client for the Tailscale API."
-authors = ["Franck Nijhof <opensource@frenck.dev>"]
+authors = ["Franck Nijhof <opensource@frenck.dev>", "Clay Rosenthal <clayrosenthal@gmail.com>"]
```
Comment (frenck) [inline pyproject.toml:5]: This is not for contributors, please revert

File: pyproject.toml [hunk 5]
```diff
@@ -155,6 +155,14 @@ ignore-imports = true
 [tool.pylint.FORMAT]
 max-line-length=88
 
+[tool.pyright]
```
Comment (frenck) [inline pyproject.toml:158]: Unrelated change

File: src/tailscale/models.py [hunk 6]
```diff
@@ -35,6 +35,7 @@ class Device(BaseModel):
 
     addresses: List[str]
     device_id: str = Field(..., alias="id")
+    node_id: str = Field(..., alias="nodeId")
```
Comment (frenck) [inline src/tailscale/models.py:38]: Unrelated change

File: src/tailscale/models.py [hunk 7]
```diff
@@ -51,9 +52,12 @@ class Device(BaseModel):
     machine_key: str = Field(..., alias="machineKey")
     node_key: str = Field(..., alias="nodeKey")
     blocks_incoming_connections: bool = Field(..., alias="blocksIncomingConnections")
+    # not included in default response
     enabled_routes: List[str] = Field(alias="enabledRoutes", default_factory=list)
     advertised_routes: List[str] = Field(alias="advertisedRoutes", default_factory=list)
-    client_connectivity: ClientConnectivity = Field(alias="clientConnectivity")
+    client_connectivity: Optional[ClientConnectivity] = Field(
```
Comment (frenck) [inline src/tailscale/models.py:58]: Unrelated change

File: src/tailscale/tailscale.py [hunk 8]
```diff
@@ -93,28 +206,51 @@ async def _request(
...
             raise TailscaleConnectionError(
                 "Error occurred while communicating with the Tailscale API"
             ) from exception
 
-        response_data: dict[str, Any] = await response.json(content_type=None)
+        response_data: Dict[str, Any] = await response.json(content_type=None)
         return response_data
 
-    async def devices(self) -> dict[str, Device]:
+    async def devices(self, all_fields: bool = True) -> Dict[str, Device]:
```
Comment (frenck) [inline src/tailscale/tailscale.py:225]: Unrelated change

File: src/tailscale/tailscale.py [hunk 9]
```diff
@@ -93,28 +206,51 @@ async def _request(
...
         Returns:
             Returns a dictionary of Tailscale devices.
         """
-        data = await self._request(f"tailnet/{self.tailnet}/devices?fields=all")
+        data = await self._get(
+            f"tailnet/{self.tailnet}/devices{'?fields=all' if all_fields else ''}"
+        )
         return Devices.parse_obj(data).devices
 
+    async def device(self, device_id: str, all_fields: bool = True) -> Device:
```
Comment (frenck) [inline src/tailscale/tailscale.py:239]: Unrelated change

**Review by frenck — CHANGES_REQUESTED** [review]
Please don't make changes that are not related to the core/goal
Of the pr.

This is the same review comment as the previous PR.

**Comment by clayrosenthal** [comment]
I did the first round from memory, thanks for clarifying the extras. Tried my best to cleanup the commit history, but I still think you would want to squash it

**Comment by clayrosenthal** [comment]
hey @frenck, do you have any other feedback or revisions you'd like me to make? Is it still too large?

**Pull request closed without merging**